### PR TITLE
bpo-44874: deprecate Py_TRASHCAN_SAFE_BEGIN and Py_TRASHCAN_SAFE_END

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue44874

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst [hunk 1]
```diff
@@ -0,0 +1,38 @@
+Deprecate the old trashcan macros (``Py_TRASHCAN_SAFE_BEGIN``/``Py_TRASHCAN_SAFE_END``). They should be replaced by the new macros ``Py_TRASHCAN_BEGIN`` and ``Py_TRASHCAN_END``.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst:1]: Few people read the Changelog. I suggest to move these explanations to What's New in Python 3.11 > C API > Porting to Python 3.11:
https://docs.python.org/dev/whatsnew/3.11.html#id3

File: Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+       ...
+       Py_TRASHCAN_END
+   }
+
+Note that ``Py_TRASHCAN_BEGIN`` has a second argument which
+ should be the deallocation function it is in.
+
+To support older python versions in the same codebase, you
+can define the following macros and use them throughout
+the code (credit: these were copied from the ``mypy`` codebase)::
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst]: It seems like mypy is released under the MIT license: https://github.com/python/mypy/blob/master/LICENSE

File: Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+   #if PY_MAJOR_VERSION >= 3 && PY_MINOR_VERSION >= 8
+   #define CPy_TRASHCAN_BEGIN(op, dealloc) Py_TRASHCAN_BEGIN(op, dealloc)
+   #define CPy_TRASHCAN_END(op) Py_TRASHCAN_END
+   #else
+   #define CPy_TRASHCAN_BEGIN(op, dealloc) Py_TRASHCAN_SAFE_BEGIN(op)
+   #define CPy_TRASHCAN_END(op) Py_TRASHCAN_SAFE_END(op)
+   #endif
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst]: For readability:

```suggestion
   #if PY_MAJOR_VERSION >= 3 && PY_MINOR_VERSION >= 8
   #  define CPy_TRASHCAN_BEGIN(op, dealloc) Py_TRASHCAN_BEGIN(op, dealloc)
   #  define CPy_TRASHCAN_END(op) Py_TRASHCAN_END
   #else
   #  define CPy_TRASHCAN_BEGIN(op, dealloc) Py_TRASHCAN_SAFE_BEGIN(op)
   #  define CPy_TRASHCAN_END(op) Py_TRASHCAN_SAFE_END(op)
   #endif
```

**Review by vstinner — COMMENTED** [review]
Thanks for the nice explanation on how to port existing code!

File: Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst (hunk 2 above)
Comment (ambv) [reply Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst]: How does this impact the example?

File: Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst (hunk 1 above)
Comment (ambv) [reply Misc/NEWS.d/next/Core and Builtins/2021-08-09-19-05-20.bpo-44874.oOcfU4.rst:1]: I'll add that.

**Pull request merged**